pkg/storage/etcd3/store.go

```diff
@@ -203,8 +211,45 @@ func (s *store) List(ctx context.Context, key, resourceVersion string, filter st
 	return s.versioner.UpdateList(listObj, uint64(getResp.Header.Revision))
 }
 
-func (s *store) GuaranteedUpdate(ctx context.Context, key string, ptrToType runtime.Object, ignoreNotFound bool, tryUpdate storage.UpdateFunc) error {
-	panic("TODO: unimplemented")
+func (s *store) GuaranteedUpdate(ctx context.Context, key string, out runtime.Object, ignoreNotFound bool, tryUpdate storage.UpdateFunc) error {
+	v, err := conversion.EnforcePtr(out)
+	if err != nil {
+		panic("need ptr to type")
+	}
+	key = keyWithPrefix(s.pathPrefix, key)
+	for {
+		origState, err := s.getCurrentObjState(ctx, key, v, ignoreNotFound)
+		if err != nil {
+			return err
+		}
+
+		ret, err := s.getUserUpdateRet(origState, tryUpdate)
+		if err != nil {
+			return err
+		}
+
+		data, err := runtime.Encode(s.codec, ret)
+		if err != nil {
+			return err
+		}
+		if bytes.Equal(data, origState.data) {
+			return decode(s.codec, s.versioner, origState.data, out, origState.rev)
+		}
+
+		txnResp, err := s.client.KV.Txn(ctx).If(
+			sameRevision(key, origState.rev),
+		).Then(
+			clientv3.OpPut(key, string(data)),
+		).Commit()
```

Comment:
we can save a roundtrip time if we get the current state in the else statement. Or we have to do get again at line 221 for the next round. (add a tood at least)